pkg/proxy/middleware.go

```diff
@@ -0,0 +1,255 @@
+	}
+	statusCodeIndex := make(map[int]struct{}, len(optRetryStatusCodes))
+	for _, c := range optRetryStatusCodes {
+		statusCodeIndex[c] = struct{}{}
+	}
+	return &RetryMiddleware{
+		nextHandler:      other,
+		maxRetries:       maxRetries,
+		retryStatusCodes: statusCodeIndex,
+		rSource:          rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+type xResponseWriter interface {
+	http.ResponseWriter
+	discardedResponse() bool
+	CapturedStatusCode() int
+}
+type xBodyCapturer interface {
+	io.ReadCloser
+	Capture()
+}
+
+func (r RetryMiddleware) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	lazyBody := newLazyBodyCapturer(request.Body)
+	request.Body = lazyBody
+	for i := 0; ; i++ {
+		discardErrResp := i < r.maxRetries && request.Context().Err() == nil
+		capturedResp := newResponseWriterDelegator(writer, r.isRetryableStatusCode, discardErrResp)
+		// call next handler in chain
+		reqClone := request.Clone(request.Context()) // also copies the reference to the lazy body capturer
+		r.nextHandler.ServeHTTP(capturedResp, reqClone)
+
+		if !capturedResp.discardedResponse() || // max retries reached or context error
+			!r.isRetryableStatusCode(capturedResp.CapturedStatusCode()) {
+			break
+		}
+		// setup for retry
+		lazyBody.Capture()
+
+		totalRetries.Inc()
+		// exponential backoff
+		jitter := time.Duration(r.rSource.Intn(50))
+		time.Sleep(time.Millisecond*time.Duration(1<<uint(i)) + jitter)
+	}
+}
+
+func (r RetryMiddleware) isRetryableStatusCode(status int) bool {
+	_, ok := r.retryStatusCodes[status]
+	return ok
+}
+
+var (
+	_ http.Flusher       = &responseWriterDelegator{}
+	_ io.ReaderFrom      = &xResponseWriterDelegator{}
+	_ statusCodeCapturer = &responseWriterDelegator{}
+)
+
+// responseWriterDelegator represents a wrapper around http.ResponseWriter that provides additional
+// functionalities for handling response writing. Depending on the status code and discard settings,
+// the header + content on write is skipped so that it can be re-used on retry.
+type responseWriterDelegator struct {
+	http.ResponseWriter
+	headerBuf   http.Header
+	wroteHeader bool
+	statusCode  int
+	// always writes to responseWriter when false
+	discardErrResp        bool
+	isRetryableStatusCode func(status int) bool
+}
+
+// newResponseWriterDelegator constructor
+func newResponseWriterDelegator(writer http.ResponseWriter, isRetryableStatusCode func(status int) bool, discardErrResp bool) xResponseWriter {
+	d := &responseWriterDelegator{
+		isRetryableStatusCode: isRetryableStatusCode,
+		ResponseWriter:        writer,
+		headerBuf:             make(http.Header),
+		discardErrResp:        discardErrResp, // abort early on timeout, context cancel
+	}
+	if _, ok := writer.(io.ReaderFrom); ok {
+		return &xResponseWriterDelegator{responseWriterDelegator: d}
+	}
+	return d
+}
+
+func (r *responseWriterDelegator) discardedResponse() bool {
```

Comment:
Seems like this method could go away in favor of just checking `r.discardErrResp`?

Reply:
I need this to satisfy the response type in `newResponseWriterDelegator() xResponseWriter`. The source writer either implements `ReaderFrom` or not. So does the returned implementation.